src/mpid/ch4/shm/ipc/xpmem/xpmem_post.h

```diff
@@ -75,9 +75,11 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_attach_mem(int node_rank,
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_ATTACH_MEM);
 
 #ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+    MPIDI_XPMEMI_seg_t *seg_ptr = (MPIDI_XPMEMI_seg_t *) mem_seg->seg_ptr;
```

Comment:
If we move the attach and close functions to a `.c` file, we can hide all of the `XPMEMI` pieces.  The pre and post headers are exposed to the IPC layer.

Reply:
OK

Reply:
It might look messy that we expose noninline functions in both `xpmem_noinline.h` (for init/finalize) and `xpmem_post.h` (for attach_mem/close_mem).

Reply:
Except for the attribute query function, nothing else is performance critical.  So it doesn't make sense for them to be inlined at all.

Reply:
I think my question is, should we move all noninline functions into a single header file (e.g., xpmem_noinline)?

Reply:
Function declarations for exposed functions typically go in `_post.h`.  But yes, a single header.

My suggestion is:

1. Move all function declarations to `xpmem_post.h`.
2. Add the function definition for the attribute query function as a static inline to `xpmem_post.h`.
3. Add a new file `xpmem_foo.c` that has the function definitions for the remaining functions.